google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java

```diff
@@ -119,18 +123,46 @@ public static Builder newBuilder() {
   @AutoValue.Builder
   public abstract static class Builder {
 
+    /** Interface to allow tests to mock service account credentials creation. */
+    @VisibleForTesting
+    interface ServiceAccountCredentialsFactory {
+      ServiceAccountCredentials fromStream(InputStream inputStream) throws IOException;
+    }
     /**
      * The default file name for the properties configuration file. This does <em>not</em> include a
      * path.
      */
     public static final String DEFAULT_PROPERTIES_CONFIG_FILE_NAME = "ads.properties";
+
+    /** Required scope of credentials for the Google Ads API. */
+    private static final String GOOGLE_ADS_API_SCOPE = "https://www.googleapis.com/auth/adwords";
+
+    /** Keys for creating installed app/web flow credentials. */
+    private static final ImmutableSet<ConfigPropertyKey> INSTALLED_APP_OAUTH_KEYS =
+        ImmutableSet.of(
+            ConfigPropertyKey.CLIENT_ID,
+            ConfigPropertyKey.CLIENT_SECRET,
+            ConfigPropertyKey.REFRESH_TOKEN);
+
+    /** Keys for creating service account credentials. */
+    private static final ImmutableSet<ConfigPropertyKey> SERVICE_ACCOUNT_OAUTH_KEYS =
+        ImmutableSet.of(
+            ConfigPropertyKey.SERVICE_ACCOUNT_SECRETS_PATH, ConfigPropertyKey.SERVICE_ACCOUNT_USER);
+
     /**
      * Allows injecting a different default configuration file, instead of using {@value
      * #DEFAULT_PROPERTIES_CONFIG_FILE_NAME}. Only for testing.
      */
     private Supplier<File> configurationFileSupplier =
         () -> new File(System.getProperty("user.home"), DEFAULT_PROPERTIES_CONFIG_FILE_NAME);
 
+    /**
+     * Allows injecting a different factory method for creating {@link ServiceAccountCredentials}
+     * from a stream. Only for testing.
+     */
+    private ServiceAccountCredentialsFactory serviceAccountCredentialsFactory =
+        inputStream -> ServiceAccountCredentials.fromStream(inputStream);
```

Comment:
I'm a bit confused why you need to override this implementation. Looking at the [javadoc](https://googleapis.dev/java/google-auth-library/latest/com/google/auth/oauth2/ServiceAccountCredentials.html#fromStream-java.io.InputStream-), it doesn't seem like this method does anything that undesirable. It would be preferable to have the unit tests follow as similar a code path as possible to prod. For instance, if the service account credentials file changes format would we not not detect that in unit tests? What do you think about refactoring the tests to provide a service account credential via `InputStream`?

Reply:
This was to workaround the fact that `ServiceAccountCredentials.fromStream` requires a valid `private_key` in the file, among other things. I found a better way, which is to create the new `mock_service_account_secrets.json` file. This contains a `private_key` that I generated as follows:

```
    KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("RSA");
    keyPairGenerator.initialize(512);
    KeyPair keyPair = keyPairGenerator.generateKeyPair();
    byte[] privateKeyBytes = keyPair.getPrivate().getEncoded();
    byte[] base64Bytes = Base64.getMimeEncoder().encode(
        privateKeyBytes);
    String encodedString = new String(base64Bytes, StandardCharsets.UTF_8);
    System.out.println(encodedString);
```